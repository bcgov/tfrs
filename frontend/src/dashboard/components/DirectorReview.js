import React from 'react'
import PropTypes from 'prop-types'

import Loading from '../../app/components/Loading'
import COMPLIANCE_REPORTING from '../../constants/routes/ComplianceReporting'
import CONFIG from '../../config'
import CREDIT_TRANSACTIONS from '../../constants/routes/CreditTransactions'
import PERMISSIONS_COMPLIANCE_REPORT from '../../constants/permissions/ComplianceReport'
import PERMISSIONS_CREDIT_TRANSACTIONS from '../../constants/permissions/CreditTransactions'
import { useNavigate } from 'react-router'

const DirectorReview = (props) => {
  const { isGettingDashboard: fetchingDashboard, supplementalItems: supplementalReports } = props.complianceReports
  const navigate = useNavigate()
  const { isFinding: fetchingCreditTransfers, items: creditTransfers } =
    props.creditTransfers

  const awaitingReview = {
    complianceAndExclusionReports: 0,
    creditTransfers: 0,
    exclusionReports: 0,
    part3Awards: 0,
    total: 0
  }

  if (
    CONFIG.COMPLIANCE_REPORTING.ENABLED &&
    typeof props.loggedInUser.hasPermission === 'function' &&
    props.loggedInUser.hasPermission(PERMISSIONS_COMPLIANCE_REPORT.VIEW)
  ) {
    supplementalReports &&
    supplementalReports.forEach((item) => {
      const { status } = item
      if (['Not Recommended', 'Recommended'].indexOf(status.managerStatus) >= 0 && status.directorStatus === 'Unreviewed') {
        awaitingReview.complianceAndExclusionReports += 1
      }
    })
  }

  if (
    typeof props.loggedInUser.hasPermission === 'function' &&
    props.loggedInUser.hasPermission(PERMISSIONS_CREDIT_TRANSACTIONS.VIEW)
  ) {
    creditTransfers.forEach((item) => {
      if (!item.isRescinded) {
        if (
          ['Recommended', 'Not Recommended'].indexOf(item.status.status) >= 0
        ) {
          if (['Buy', 'Sell'].indexOf(item.type.theType) >= 0) {
            awaitingReview.creditTransfers += 1
            awaitingReview.total += 1
          }

          if (['Part 3 Award'].indexOf(item.type.theType) >= 0) {
            awaitingReview.part3Awards += 1
            awaitingReview.total += 1
          }
        }
      }
    })
  }

  return (
    <div className="dashboard-fieldset director">
      <h1>Director Review</h1>

      <div>
      There are:
        <>
          <div className="content">
            {typeof props.loggedInUser.hasPermission === 'function' &&
              props.loggedInUser.hasPermission(
                PERMISSIONS_CREDIT_TRANSACTIONS.VIEW
              ) && (
                <div>
                  <span>  {fetchingCreditTransfers ? <Loading/> : `${awaitingReview.creditTransfers} `}</span>
                  <button
                    onClick={() => {
                      props.setFilter(
                        [
                          {
                            id: 'compliancePeriod',
                            value: ''
                          },
                          {
                            id: 'transactionType',
                            value: 'Credit Transfer'
                          },
                          {
                            id: 'status',
                            value: 'Reviewed'
                          }
                        ],
                        'credit-transfers'
                      )

                      return navigate(CREDIT_TRANSACTIONS.LIST)
                    }}
                    type="button"
                    data-testid="creditTransferBtn"
                  >

                  Credit transfer(s) for your review and statutory decision
                  </button>
                </div>
            )}

            {CONFIG.COMPLIANCE_REPORTING.ENABLED &&
              typeof props.loggedInUser.hasPermission === 'function' &&
              props.loggedInUser.hasPermission(
                PERMISSIONS_COMPLIANCE_REPORT.VIEW
              ) && (
                <div>
                  <span>{fetchingDashboard ? <Loading/> : awaitingReview.complianceAndExclusionReports}</span>
                  <button
                    className="btn-text"
                    onClick={() => {
                      props.setFilter(
                        [
                          {
                            id: 'current-status',
                            value: ['For Director Review']
                          }

                        ],
                        'compliance-reporting'
                      )

                      return navigate(COMPLIANCE_REPORTING.LIST, { state: { items: ['For Director Review'] } })
                    }}
                    type="button"
                  >
                   Compliance / exclusion report(s) awaiting your review
                  </button>
                </div>
            )}

            {typeof props.loggedInUser.hasPermission === 'function' &&
              props.loggedInUser.hasPermission(
                PERMISSIONS_CREDIT_TRANSACTIONS.VIEW
              ) && (
                <div>
                <span> {awaitingReview.part3Awards}</span>
                  <button
                    onClick={() => {
                      props.setFilter(
                        [
                          {
                            id: 'compliancePeriod',
                            value: ''
                          },
                          {
                            id: 'transactionType',
                            value: 'Part 3 Award'
                          },
                          {
                            id: 'status',
                            value: 'Reviewed'
                          }
                        ],
                        'credit-transfers'
                      )

                      return navigate(CREDIT_TRANSACTIONS.LIST)
                    }}
                    type="button"
                  >
                    Part 3 Award(s) awaiting your
                    review
                  </button>
                </div>
            )}
          </div>
        </>
      </div>
    </div>
  )
}

DirectorReview.defaultProps = {}

DirectorReview.propTypes = {
  complianceReports: PropTypes.shape({
    isFetching: PropTypes.bool,
    isGettingDashboard: PropTypes.bool,
    items: PropTypes.arrayOf(PropTypes.shape()),
    supplementalItems: PropTypes.arrayOf(PropTypes.shape())
  }).isRequired,
  creditTransfers: PropTypes.shape({
    isFetching: PropTypes.bool,
    isFinding: PropTypes.bool,
    items: PropTypes.arrayOf(PropTypes.shape())
  }).isRequired,
  loggedInUser: PropTypes.shape().isRequired,
  setFilter: PropTypes.func.isRequired
}

export default DirectorReview
