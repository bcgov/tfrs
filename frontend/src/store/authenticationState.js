import { loadUser } from 'redux-oidc';
import { put, takeLatest, all } from 'redux-saga/effects';
import userManager from './oidc-usermanager';
import { getLoggedInUser } from '../actions/userActions';
import CONFIG from '../config';

const LOGIN_TRIGGERING_ACTIONS = [
  'redux-oidc/USER_EXPIRED'
];

function triggerLoginFlow (store) {
  const { routing } = store.getState();

  if (routing.location &&
    routing.location.pathname !== '/authCallback') {
    return CONFIG.KEYCLOAK.CUSTOM_LOGIN ? userManager.signinSilent() : userManager.signinRedirect();
  }

  return false;
}

function * getBackendUser () {
  yield put(getLoggedInUser());
}

export default function * authenticationStateSaga (store) {
  userManager.clearStaleState();

  const { routing } = store.getState();

  if (!routing.location || routing.location.pathname !== '/authCallback') {
    loadUser(store, userManager);
  }
  yield all([
    takeLatest('redux-oidc/USER_FOUND', getBackendUser),
    takeLatest(
      action => (LOGIN_TRIGGERING_ACTIONS.includes(action.type)),
      triggerLoginFlow,
      store
    )
  ]);
}
