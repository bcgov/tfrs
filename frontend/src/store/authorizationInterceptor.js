import axios from 'axios';
import userManager from './oidc-usermanager';
import CONFIG from '../config';

function configureAxios () {
  if (CONFIG.KEYCLOAK) {
    console.log('adding token to request');

    axios.interceptors.request.use((config) => {
      const loadconfig = async () => {
        const user = await userManager.getUser();
        return {
          ...config,
          headers: {
            Authorization: `Bearer ${user.id_token}`
          }
        };
      };

      return loadconfig();
    });
  }
}

export default configureAxios;
