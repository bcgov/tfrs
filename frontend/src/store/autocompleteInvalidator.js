import { call, put, takeLatest } from 'redux-saga/effects';
import AutocompleteActionTypes from "../constants/actionTypes/Autocomplete";
import FuelCodeActionTypes from '../constants/actionTypes/FuelCodes';

const INVALIDATING_ACTIONS = [
  FuelCodeActionTypes.ADD_FUEL_CODE,
  FuelCodeActionTypes.RECEIVE_FUEL_CODE,
  FuelCodeActionTypes.RECEIVE_FUEL_CODES,
  FuelCodeActionTypes.UPDATE_FUEL_CODE,
  FuelCodeActionTypes.DELETE_FUEL_CODE
];

function * invalidateCache () {
  yield put({ type: AutocompleteActionTypes.INVALIDATE_AUTOCOMPLETE_CACHE });
}

export default function * sessionTimeoutSaga () {
  yield takeLatest(
    action => (INVALIDATING_ACTIONS.includes(action.type)),
    invalidateCache
  );
}
