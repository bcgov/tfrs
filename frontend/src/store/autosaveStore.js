import {takeEvery, all, call} from 'redux-saga/effects';
import {delay} from "redux-saga";
import axios from "axios";
import * as Routes from "../constants/routes";

function* loadAutosaveData(action) {
  const {key, resolve, reject} = action.payload;

  axios
    .get(`${Routes.BASE_URL}${Routes.AUTOSAVE_API}?key=${key}`)
    .then((response) => {
      try {
        const parsed = JSON.parse(response.data.data);
        resolve(parsed);
      } catch (e) {
        reject(e);
      }
    }).catch((error) => {
    reject(error);
  });
}

function* saveAutosaveData(action) {
  const {key, state, resolve, reject} = action.payload;

  const payload = {
    key,
    data: JSON.stringify(state)
  };

  axios
    .post(`${Routes.BASE_URL}${Routes.AUTOSAVE_API}`, payload)
    .then((response) => {
      resolve();
    }).catch((error) => {
    reject(error);
  });
}

function* clearAutosaveData(action) {
  const {resolve, reject} = action.payload;

  axios
    .post(`${Routes.BASE_URL}${Routes.AUTOSAVE_API}/clear`)
    .then((response) => {
      try {
        resolve();
      } catch (e) {
        reject(e);
      }
    }).catch((error) => {
    reject(error);
  });
}

export default function* autosaveSaga() {

  yield all([
    takeEvery('LOAD_AUTOSAVE_DATA', loadAutosaveData),
    takeEvery('SAVE_AUTOSAVE_DATA', saveAutosaveData),
    takeEvery('CLEAR_AUTOSAVE_DATA', clearAutosaveData)
  ]);
}
