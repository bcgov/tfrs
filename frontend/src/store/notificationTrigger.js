import {call, put, takeLatest} from 'redux-saga/effects';
import {delay} from 'redux-saga';
import {getNotifications, getNotificationsCount} from "../actions/notificationActions";
import UserActionTypes from "../constants/actionTypes/Users";
import NotificationActionTypes from "../constants/actionTypes/Notifications";


const TRIGGERING_ACTIONS = [
  'SERVER_INITIATED_NOTIFICATION_RELOAD',
  UserActionTypes.RECEIVE_LOGGED_IN_USER,
  NotificationActionTypes.SUCCESS_NOTIFICATIONS
];

function* fetchNotifications(store) {
  yield call(delay, 1000); //debounce

  if (store.getState().rootReducer.userRequest.isAuthenticated) {
    yield put(getNotificationsCount());
  }

  if (store.getState().rootReducer.notifications.onNotificationsPage) {
    yield put(getNotifications());
  }

}

export default function* notificationsSaga(store) {
  yield takeLatest(
    action => (TRIGGERING_ACTIONS.includes(action.type)),
    fetchNotifications,
    store
  );
}
