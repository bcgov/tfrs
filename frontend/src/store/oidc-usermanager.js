import { createUserManager } from 'redux-oidc';
import CONFIG from '../config';

const settings = {
  authority: CONFIG.KEYCLOAK.AUTHORITY,
  client_id: CONFIG.KEYCLOAK.CLIENT_ID,
  redirect_uri: CONFIG.KEYCLOAK.CALLBACK_URL,
  automaticSilentRenew: true,
  silent_redirect_uri: `${window.location.protocol}//${window.location.hostname}${window.location.port ? `:${window.location.port}` : ''}/token_renew.html`,
  response_type: 'token id_token'
};

const userManager = createUserManager(settings);

export default userManager;
