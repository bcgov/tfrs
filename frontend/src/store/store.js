import { createStore, compose, applyMiddleware, combineReducers } from 'redux';
import { reducer as toastrReducer } from 'react-redux-toastr';
import { routerReducer, routerMiddleware } from 'react-router-redux';
import { createLogger } from 'redux-logger';
import io from 'socket.io-client';
import thunk from 'redux-thunk';
import rootReducer from '../reducers/reducer';
import { getNotifications } from '../actions/notificationActions';
import { SOCKETIO_URL } from '../constants/routes';

import createOidcMiddleware, { loadUser, processSilentRenew, reducer as OIDCReducer } from 'redux-oidc';
import { persistTargetPathReducer } from '../reducers/persistTargetPathReducer';

import userManager from './oidc-usermanager';
import CONFIG from '../config';
import { getLoggedInUser } from '../actions/userActions';

import persistState from 'redux-localstorage';
import createSocketIoMiddleware from 'redux-socket.io';

const middleware = routerMiddleware(history);
const oidcMiddleware = createOidcMiddleware(userManager);

const socket = io(SOCKETIO_URL);
const socketIoMiddleware = createSocketIoMiddleware(socket, 'socketio/');

const enhancer = compose(persistState(['targetPath'], {key: 'tfrs-state'}));

const combinedReducers = (state = {}, action) => {
  const currentRoute = state.routing || {};
  return {
    toastr: toastrReducer(state.toastr, action),
    oidc: OIDCReducer(state.oidc, action),
    routing: routerReducer(state.routing, action),
    targetPath: persistTargetPathReducer(state.targetPath, {...action, currentRoute}),
    rootReducer: rootReducer(state.rootReducer, action),
  };
};

let allMiddleware = [
  thunk,
  socketIoMiddleware,
  oidcMiddleware,
  middleware
];

if (process.env.NODE_ENV !== 'production') {
  allMiddleware.push(createLogger());
}

const store = createStore(
  combinedReducers,
  applyMiddleware(...allMiddleware),
  enhancer
);

let subscriptionProcessing = false;

store.subscribe(() => {
  const state = store.getState();
  if (!subscriptionProcessing) {
    subscriptionProcessing = true;

    if (state.rootReducer.notifications.serverInitiatedReloadRequested === true) {
      store.dispatch(getNotifications());
    }

    if (CONFIG.KEYCLOAK.ENABLED) {
      if (state.oidc.user &&
        !state.rootReducer.userRequest.isFetching &&
        !state.rootReducer.userRequest.isAuthenticated) {
        store.dispatch(getLoggedInUser());
      }
    }

    subscriptionProcessing = false;
  }
});

if (CONFIG.KEYCLOAK.ENABLED) {
  loadUser(store, userManager);
  processSilentRenew();
}

export default store;
