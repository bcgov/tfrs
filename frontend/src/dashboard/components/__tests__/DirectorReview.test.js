import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import store from '../../../store/store'
import DirectorReview from '../DirectorReview'

jest.mock('../../../config', () => {
  const originalConfig = jest.requireActual('../../../config')
  return {
    ...originalConfig,
    COMPLIANCE_REPORTING: {
      ENABLED: true
    },
    SECURE_DOCUMENT_UPLOAD: {
      ENABLED: true
    },
    FUEL_CODES: {
      ENABLED: true
    },
    DEBUG: {
      ENABLED: true
    }
  }
})

describe('DirectorReview', () => {
  const mockItems = [
    {
      id: 654,
      isRescinded: false,
      status: {
        status: 'Recommended',
        analystStatus: 'Unreviewed',
        directorStatus: 'Unreviewed',
        fuelSupplierStatus: 'Submitted',
        managerStatus: 'Unreviewed'
      },
      type: 'Compliance Report',
      supplementalReports: [{
        status: {
          managerStatus: 'Recommended',
          directorStatus: 'Unreviewed'
        }
      }]
    },
    {
      id: 655,
      status: {
        analystStatus: 'Unreviewed',
        directorStatus: 'Unreviewed',
        fuelSupplierStatus: 'Submitted',
        managerStatus: 'Unreviewed'
      },
      type: 'Compliance Report',
      supplementalReports: []
    },
    {
      id: 656,
      status: {
        analystStatus: 'Unreviewed',
        directorStatus: 'Unreviewed',
        fuelSupplierStatus: 'Submitted',
        managerStatus: 'Unreviewed'
      },
      type: 'Exclusion Report',
      supplementalReports: []
    }
  ]
  test('Should render the component', () => {
    const props = {
      complianceReports: {
        isFinding: false,
        isGettingDashboard: false,
        items: []
      },
      setFilter: jest.fn(),
      creditTransfers: {
        isFetching: false,
        isFinding: false,
        items: []
      },
      loggedInUser: {}
    }
    const component = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <DirectorReview {...props} />
        </Provider>
      </BrowserRouter>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Should render the component with complianceReports and creditTransfers', () => {
    const props = {
      complianceReports: {
        isFinding: false,
        isGettingDashboard: false,
        supplementalItems: mockItems
      },
      setFilter: jest.fn(),
      creditTransfers: {
        isFetching: false,
        isFinding: false,
        items: mockItems
      },
      loggedInUser: {
        hasPermission: jest.fn().mockImplementation(() => true)
      }
    }
    const component = renderer.create(
      <BrowserRouter>
        <Provider store={store}>
          <DirectorReview {...props} />
        </Provider>
      </BrowserRouter>
    )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
