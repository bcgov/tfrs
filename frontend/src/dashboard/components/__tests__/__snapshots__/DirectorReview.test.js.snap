// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectorReview Should render the component 1`] = `
<div
  className="dashboard-fieldset director"
>
  <h1>
    Director Review
  </h1>
  <div>
    There are:
    <div
      className="content"
    />
  </div>
</div>
`;

exports[`DirectorReview Should render the component with complianceReports and creditTransfers 1`] = `
<div
  className="dashboard-fieldset director"
>
  <h1>
    Director Review
  </h1>
  <div>
    There are:
    <div
      className="content"
    >
      <div>
        <span>
            
          0 
        </span>
        <button
          data-testid="creditTransferBtn"
          onClick={[Function]}
          type="button"
        >
          Credit transfer(s) for your review and statutory decision
        </button>
      </div>
      <div>
        <span>
          0
        </span>
        <button
          className="btn-text"
          onClick={[Function]}
          type="button"
        >
          Compliance / exclusion report(s) awaiting your review
        </button>
      </div>
      <div>
        <span>
           
          0
        </span>
        <button
          onClick={[Function]}
          type="button"
        >
          Part 3 Award(s) awaiting your review
        </button>
      </div>
    </div>
  </div>
</div>
`;
